### PR TITLE
refactor(core): Adapt the usage of routing_algorithm_id in routing and payments core for v2

### PR DESCRIPTION
**Pull request by Aprabhat19**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
In this PR we adapt the usage of routing_algorithm_id as well as payout_routing_algorithm_id in routing and payments core for v2. Also refactored the routing_algorithm list, to list all the algorithms(history) on the basis of a single  profile.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
 
 - List all the routing rules for under a profile 
```
curl --location 'http://localhost:8080/v2/profile/pro_VirB3AxB3qItFTnLsuA2/routing_algorithm' \
--header 'api-key: dev_jsssWAg0XyxGK0ycPunkFjlItYNs30X1IVOr1MIF7lwrDQgBBk7npwdkZB4t7EQd'
```
## Checklist

- [ ] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/diesel_models/src/business_profile.rs [hunk 1]
```diff
@@ -521,6 +524,9 @@ pub struct BusinessProfileNew {
     pub redirect_to_merchant_with_http_post: bool,
     pub webhook_details: Option<pii::SecretSerdeValue>,
     pub metadata: Option<pii::SecretSerdeValue>,
+    pub intent_fulfillment_time: Option<i64>,
+    pub frm_routing_algorithm: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/diesel_models/src/business_profile.rs:528]: remove this or make this into `frm_routing_algorithm_id`

File: crates/diesel_models/src/business_profile.rs [hunk 2]
```diff
@@ -555,6 +561,8 @@ pub struct BusinessProfileUpdateInternal {
     pub redirect_to_merchant_with_http_post: Option<bool>,
     pub webhook_details: Option<pii::SecretSerdeValue>,
     pub metadata: Option<pii::SecretSerdeValue>,
+    pub intent_fulfillment_time: Option<i64>,
+    pub frm_routing_algorithm: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/diesel_models/src/business_profile.rs:565]: remove this or make this into frm_routing_algorithm_id

File: crates/diesel_models/src/business_profile.rs [hunk 3]
```diff
@@ -588,6 +596,8 @@ pub enum BusinessProfileUpdate {
         redirect_to_merchant_with_http_post: Option<bool>,
         webhook_details: Option<pii::SecretSerdeValue>,
         metadata: Option<pii::SecretSerdeValue>,
+        intent_fulfillment_time: Option<i64>,
+        frm_routing_algorithm: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/diesel_models/src/business_profile.rs:600]: remove this or make this into frm_routing_algorithm_id

File: crates/hyperswitch_domain_models/src/business_profile.rs [hunk 4]
```diff
@@ -430,6 +432,9 @@ pub enum BusinessProfileUpdate {
         redirect_to_merchant_with_http_post: Option<bool>,
         webhook_details: Option<pii::SecretSerdeValue>,
         metadata: Option<pii::SecretSerdeValue>,
+        intent_fulfillment_time: Option<i64>,
+        frm_routing_algorithm: Option<serde_json::Value>,
```
Comment (jarnura) [inline crates/hyperswitch_domain_models/src/business_profile.rs:436]: remove this or make this into frm_routing_algorithm_id

File: crates/router/src/core/admin.rs [hunk 5]
```diff
@@ -3477,6 +3481,176 @@ pub async fn update_business_profile(
...
+        .as_ref()
+        .map(|webhook_details| {
+            webhook_details.encode_to_value().change_context(
+                errors::ApiErrorResponse::InvalidDataValue {
+                    field_name: "webhook details",
+                },
+            )
+        })
+        .transpose()?
+        .map(Secret::new);
```
Comment (jarnura) [inline crates/router/src/core/admin.rs:3533]: Move this to impl based function approach

Comment (jarnura) [reply crates/router/src/core/admin.rs:3533]: Below implementation which is not dependent others and just need to be parsed or convert to secret, follow impl based approach

File: crates/router/src/core/payments.rs [hunk 6]
```diff
@@ -3918,34 +3918,72 @@ pub async fn route_connector_v1<F>(
...
-        .transpose()
-        .change_context(errors::ApiErrorResponse::InternalServerError)
-        .attach_printable("Could not decode merchant routing algorithm ref")?
-        .unwrap_or_default();
+    #[cfg(all(
+        any(feature = "v1", feature = "v2"),
+        not(any(feature = "routing_v2", feature = "business_profile_v2"))
+    ))]
+    let (profile_id, connectors) = {
+        #[allow(unused_variables)]
```
Comment (jarnura) [inline crates/router/src/core/payments.rs:3957]: remove this and handle using underscore

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/admin.rs [hunk 7]
```diff
@@ -3477,6 +3481,176 @@ pub async fn update_business_profile(
...
+    request: api::BusinessProfileUpdate,
+) -> RouterResponse<api::BusinessProfileResponse> {
+    let db = state.store.as_ref();
+    let business_profile = db
+        .find_business_profile_by_profile_id(profile_id)
+        .await
+        .to_not_found_response(errors::ApiErrorResponse::BusinessProfileNotFound {
+            id: profile_id.to_owned(),
+        })?;
+    let key_store = db
```
Comment (jarnura) [inline crates/router/src/core/admin.rs:3498]: Merchant Key store is not required for v2 instead we should use the secret manager approach

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/types/api/admin.rs [hunk 8]
```diff
@@ -163,7 +166,68 @@ impl ForeignTryFrom<domain::BusinessProfile> for BusinessProfileResponse {
...
+            merchant_id: item.merchant_id,
+            profile_id: item.profile_id,
+            profile_name: item.profile_name,
+            return_url: item.return_url,
+            enable_payment_response_hash: item.enable_payment_response_hash,
+            payment_response_hash_key: item.payment_response_hash_key,
+            redirect_to_merchant_with_http_post: item.redirect_to_merchant_with_http_post,
+            webhook_details: item.webhook_details.map(ForeignInto::foreign_into),
+            metadata: item.metadata,
+            routing_algorithm: todo!(),
```
Comment (Narayanbhat166) [inline crates/router/src/types/api/admin.rs:200]: maybe add a short description as to what has to be done here and other places

File: crates/diesel_models/src/business_profile.rs [hunk 9]
```diff
@@ -231,6 +231,7 @@ pub struct BusinessProfile {
     pub redirect_to_merchant_with_http_post: bool,
     pub webhook_details: Option<WebhookDetails>,
     pub metadata: Option<pii::SecretSerdeValue>,
+    pub intent_fulfillment_time: Option<i64>,
```
Comment (jarnura) [inline crates/diesel_models/src/business_profile.rs:234]: if the field already present in v1 don't change the order, lets keep the same field order

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/admin.rs [hunk 10]
```diff
@@ -3606,6 +3607,91 @@ pub async fn update_business_profile(
...
+        storage_impl::redis::cache::CacheKind::Routing(
+            format!(
+                "routing_config_{}_{profile_id}",
+                merchant_id.get_string_repr()
+            )
+            .into(),
+        )
+    }
+
+    pub async fn update_business_profile_active_algorithm_id(
```
Comment (jarnura) [inline crates/router/src/core/admin.rs:3652]: update_business_profile_and_invalidate_routing_config_for_active_algorithm_id_update

File: crates/router/src/routes/app.rs [hunk 11]
```diff
@@ -1493,23 +1498,36 @@ impl BusinessProfile {
...
+                                routing::routing_unlink_config(
+                                    state,
+                                    req,
+                                    path,
+                                    &TransactionType::Payment,
+                                )
+                            },
+                        )),
+                    )
+                    .service(web::resource("/routing_algorithm").route(web::get().to(
```
Comment (Narayanbhat166) [inline crates/router/src/routes/app.rs:1513]: Do not call the function directly here, define a route in routes and then call that here

File: crates/router/src/core/payments.rs [hunk 12]
```diff
@@ -3919,35 +3919,59 @@ pub async fn route_connector_v1<F>(
...
-        TransactionData::Payment(payment_data) => (
-            payment_data.payment_intent.profile_id.clone(),
-            business_profile.routing_algorithm.clone(),
-        ),
-        #[cfg(feature = "payouts")]
-        TransactionData::Payout(payout_data) => (
-            Some(payout_data.payout_attempt.profile_id.clone()),
-            business_profile.payout_routing_algorithm.clone(),
-        ),
+    #[cfg(all(
```
Comment (jarnura) [inline crates/router/src/core/payments.rs:3922]: get function based approach instead using a inline cfg's

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/routes/app.rs [hunk 13]
```diff
@@ -1493,23 +1498,36 @@ impl BusinessProfile {
...
+                        },
+                    ))),
             )
     }
 }
 #[cfg(all(
     feature = "olap",
     any(feature = "v1", feature = "v2"),
-    not(feature = "routing_v2")
+    not(any(feature = "routing_v2", feature = "business_profile_v2"))
```
Comment (Narayanbhat166) [inline crates/router/src/routes/app.rs:1530]: why is this any? This should be all

File: crates/router/src/core/routing.rs [hunk 14]
```diff
@@ -499,7 +494,11 @@ pub async fn retrieve_routing_config(
     Ok(service_api::ApplicationResponse::Json(response))
 }
 
-#[cfg(all(feature = "v2", feature = "routing_v2"))]
+#[cfg(all(
+    feature = "v2",
+    feature = "routing_v2",
+    feature = "business_profile_v2"
+))]
 pub async fn unlink_routing_config(
```
Comment (jarnura) [inline crates/router/src/core/routing.rs:502]:
```suggestion
pub async fn unlink_routing_algorithm_from_profile(
```

File: crates/router/src/core/routing.rs [hunk 15]
```diff
@@ -521,44 +520,29 @@ pub async fn unlink_routing_config(
     .await?
     .get_required_value("BusinessProfile")?;
```
Comment (jarnura) [inline crates/router/src/core/routing.rs:521]: we can move the wrapper new function here

**Review by jarnura — DISMISSED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by lsampras — APPROVED** [review]

**Review by prajjwalkumar17 — APPROVED** [review]

**Pull request merged**